# **Model Performance Comparison**

Five different machine learning experiments were performed to identify the most suitable model for predicting online shopper purchase intention. Each experiment was evaluated using Accuracy, Precision, Recall, F1-Score, ROC-AUC, and Overfitting. Since the dataset is imbalanced, model selection was based on multiple evaluation metrics rather than accuracy alone.

## Comparision Table 

In [11]:
import pandas as pd

comparison = pd.DataFrame({

    "Experiment": [
        "Exp 1",
        "Exp 2",
        "Exp 3",
        "Exp 4 (DT)",
        "Exp 4 (RF)"
    ],

    "Model": [
        "Logistic Regression",
        "Logistic Regression (Log1p)",
        "Logistic Regression (Log1p + Class Weight)",
        "Decision Tree (GridSearchCV)",
        "Random Forest (RandomizedSearchCV)"
    ],

    "Test Accuracy": [
        0.89,
        0.9029,
        0.87,
        0.85,      
        0.89       
    ],

    "Precision": [
        0.77,
        0.72,
        0.56,
        0.52,      
        0.64      
    ],

    "Recall": [
        0.42,
        0.61,
        0.81,
        0.81,     
        0.75      
    ],

    "F1-Score": [
        0.54,
        0.66,
        0.66,
        0.63,     
        0.69       
    ],

    "Test ROC-AUC": [
        0.902,
        0.926,
        0.9293,
        0.9194,    
        0.9285
    ],

    "Overfitting": [
        "No",
        "No",
        "No",
        "Reduced after tuning",
        "Reduced after tuning"
    ]

})

comparison

,Experiment,Model,Test Accuracy,Precision,Recall,F1-Score,Test ROC-AUC,Overfitting
0,Exp 1,Logistic Regression,0.8900,0.77,0.42,0.54,0.9020,No
1,Exp 2,Logistic Regression (Log1p),0.9029,0.72,0.61,0.66,0.9260,No
2,Exp 3,Logistic Regression (Log1p + Class Weight),0.8700,0.56,0.81,0.66,0.9293,No
3,Exp 4 (DT),Decision Tree (GridSearchCV),0.8500,0.52,0.81,0.63,0.9194,Reduced after tuning
4,Exp 4 (RF),Random Forest (RandomizedSearchCV),0.8900,0.64,0.75,0.69,0.9285,Reduced after tuning


In [12]:
comparison.style.highlight_max(
    subset=["Test Accuracy","Precision","Recall","F1-Score","Test ROC-AUC"],
    color="lightgreen"
)

,Experiment,Model,Test Accuracy,Precision,Recall,F1-Score,Test ROC-AUC,Overfitting
0,Exp 1,Logistic Regression,0.890000,0.770000,0.420000,0.540000,0.902000,No
1,Exp 2,Logistic Regression (Log1p),0.902900,0.720000,0.610000,0.660000,0.926000,No
2,Exp 3,Logistic Regression (Log1p + Class Weight),0.870000,0.560000,0.810000,0.660000,0.929300,No
3,Exp 4 (DT),Decision Tree (GridSearchCV),0.850000,0.520000,0.810000,0.630000,0.919400,Reduced after tuning
4,Exp 4 (RF),Random Forest (RandomizedSearchCV),0.890000,0.640000,0.750000,0.690000,0.928500,Reduced after tuning


### Based on overall predictive performance, the models are ranked as follows:
**RANK 1. Exp4 – Random Forest (Best Overall)**
- Highest F1 (0.69)
- Excellent ROC-AUC (0.9285)
- Good Precision + Good Recall
- Strongest overall predictive performance

**RANK 2. Exp2 – Logistic Regression + Log1p**
- Best Logistic Regression
- Highest Accuracy (0.9029)
- Interpretable and balanced

**RANK 3. Exp3 – Logistic Regression + Log1p + Class Weight**
- Best when Recall is the priority
- Highest ROC-AUC (0.9293)
- Recall improved from 0.61 to 0.81, meaning the model detected more purchasing customers. However, this improvement came with a reduction in Accuracy and Precision, resulting in more false positive predictions.

- This model is suitable when missing a potential buyer is more costly than contacting extra non-buyers.

**RANK 4. Exp4 – Tuned Decision Tree**
- High Recall (0.81) but model still produced the lowest Accuracy (85%), lowest Precision (0.52), and lower F1-score (0.63). Therefore, it was outperformed by both Logistic Regression and Random Forest.

**RANK 5. Exp1 – Baseline Logistic Regression** 
- This model serves as a benchmark. While it achieved the highest Precision (0.77), its Recall (0.42) was very low, meaning it missed a large number of actual buyers.

# Conculsion

Although Logistic Regression with Log1p Transformation achieved the highest accuracy (90.29%), accuracy alone was not considered sufficient because the dataset is imbalanced. Greater emphasis was placed on Recall, F1-Score, ROC-AUC, and model generalization.

"I would choose the tuned Random Forest because it provides the best overall balance between Accuracy, Precision, Recall, F1-Score, ROC-AUC, and Generalization.

It has the highest F1-score (0.69), indicating the best trade-off between precision and recall, and excellent ROC-AUC (0.9286), showing excellent class separation. 

Since customer purchase prediction requires both identifying buyers accurately and minimizing missed buyers, Random Forest is the most suitable production model."